src/main/java/com/google/sps/servlets/FreeTimeServlet.java

```diff
@@ -0,0 +1,209 @@
+      }
+      freeTimeUtility.addEvent(eventStart, eventEnd);
+    }
+
+    int wordCount = getWordCount(googleCredential);
+    int averageReadingSpeed = 50;
+    int minutesToRead = (int) Math.ceil((double) wordCount / averageReadingSpeed);
+    long timeNeeded = minutesToRead * TimeUnit.MINUTES.toMillis(1);
+    List<DateInterval> potentialTimes = getPotentialTimes(freeTimeUtility, timeNeeded);
+
+    PlanMailResponse planMailResponse =
+        new PlanMailResponse(wordCount, averageReadingSpeed, minutesToRead, potentialTimes);
+    // Convert event list to JSON and print to response
+    JsonUtility.sendJson(response, planMailResponse);
+  }
+
+  private List<DateInterval> getPotentialTimes(FreeTimeUtility freeTimeUtility, long timeNeeded) {
+    List<DateInterval> workFreeInterval = freeTimeUtility.getWorkFreeInterval();
+    List<DateInterval> potentialEventTimes = new ArrayList<>();
+    long remainingTime = timeNeeded;
+    for (DateInterval interval : workFreeInterval) {
+      Date potentialEnd = new Date(interval.getStart().getTime() + remainingTime);
+      if (interval.getEnd().before(potentialEnd)) {
+        remainingTime -= (interval.getEnd().getTime() - interval.getStart().getTime());
+        potentialEventTimes.add(interval);
+      } else {
+        potentialEventTimes.add(new DateInterval(interval.getStart(), potentialEnd));
+        break;
+      }
+    }
+    return potentialEventTimes;
+  }
+
+  private int getWordCount(Credential googleCredential) {
+    GmailClient gmailClient = gmailClientFactory.getGmailClient(googleCredential);
+    GmailClient.MessageFormat messageFormat = GmailClient.MessageFormat.FULL;
+    int numberDays = 7;
+    int wordCount = 0;
+    List<Message> unreadMessages = new ArrayList<>();
+    try {
+      unreadMessages = gmailClient.getUnreadEmailsFromNDays(messageFormat, numberDays);
+    } catch (IOException e) {
+      System.out.println(e);
+    }
+    for (Message message : unreadMessages) {
+      try {
+        wordCount += getMessageSize(message);
+      } catch (MessagingException | IOException e) {
+        System.out.println(e);
+      }
+    }
+    return wordCount;
+  }
+
+  /**
+   * Get the events in the user's calendars
+   *
+   * @param calendarClient either a mock CalendarClient or a calendarClient with a valid credential
+   * @param timeMin the minimum time to start looking for events
+   * @param timeMax the maximum time to look for events
+   * @return List of Events from all of the user's calendars
+   * @throws IOException if an issue occurs in the method
+   */
+  private List<Event> getEvents(CalendarClient calendarClient, Date timeMin, Date timeMax)
+      throws IOException {
+    List<CalendarListEntry> calendarList = calendarClient.getCalendarList();
+    List<Event> events = new ArrayList<>();
+    for (CalendarListEntry calendar : calendarList) {
+      events.addAll(calendarClient.getUpcomingEvents(calendar, timeMin, timeMax));
+    }
```

Comment:
Would it be possible to do `calendarList.stream().map(calendar -> events.addAll(calendarClient.getUpcomingEvents(calendar, timeMin, timeMax)))`?

Reply:
Not exactly like that, handling of exception requires to increase the expression when using this method. I think it might get more complicated instead of easier

Reply:
What about using `Stream.flatMap` to flatten the (conceptually) nested list of events in the list of calendars? I'll leave the details as an exercise for you.

Alternatively, you could use the idea that @tttzach suggested, but replace `Stream.map` with `Stream.forEach`.